File: sdks/csharp/Renamed/AsyncJob.cs
using Renamed.Sdk.Exceptions;
using Renamed.Sdk.Logging;
using Renamed.Sdk.Models;

namespace Renamed.Sdk;

/// <summary>
/// Async job handle for long-running operations like PDF split.
/// Provides methods to poll for status and wait for completion.
/// </summary>
/// <typeparam name="TResult">The type of result expected when the job completes.</typeparam>
public sealed class AsyncJob<TResult> where TResult : class
{
    private readonly RenamedClient _client;
    private readonly string _statusUrl;
    private readonly TimeSpan _pollInterval;
    private readonly int _maxAttempts;
    private readonly IRenamedLogger? _logger;

    internal AsyncJob(
        RenamedClient client,
        string statusUrl,
        IRenamedLogger? logger = null,
        TimeSpan? pollInterval = null,
        int? maxAttempts = null)
    {
        _client = client ?? throw new ArgumentNullException(nameof(client));
        _statusUrl = statusUrl ?? throw new ArgumentNullException(nameof(statusUrl));
        _logger = logger;
        _pollInterval = pollInterval ?? TimeSpan.FromSeconds(2);
        _maxAttempts = maxAttempts ?? 150; // 5 minutes at 2s intervals
    }

    /// <summary>
    /// Gets the current job status.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The current job status response.</returns>
    public Task<JobStatusResponse> GetStatusAsync(CancellationToken cancellationToken = default)
    {
        return _client.RequestAsync<JobStatusResponse>(_statusUrl, HttpMethod.Get, cancellationToken: cancellationToken);
    }

    /// <summary>
    /// Waits for the job to complete, polling at regular intervals.
    /// </summary>
    /// <param name="onProgress">Optional callback invoked when status is updated.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The job result when completed.</returns>
    /// <exception cref="JobException">Thrown when the job fails or times out.</exception>
    public async Task<TResult> WaitAsync(
        Action<JobStatusResponse>? onProgress = null,
        CancellationToken cancellationToken = default)
    {
        var attempts = 0;
        string? lastLoggedStatus = null;
        int? lastLoggedProgress = null;

        while (attempts < _maxAttempts)
        {
            cancellationToken.ThrowIfCancellationRequested();

            var status = await GetStatusAsync(cancellationToken).ConfigureAwait(false);

            // Log job status changes and progress updates
            LogJobStatus(status, ref lastLoggedStatus, ref lastLoggedProgress);

            onProgress?.Invoke(status);

            if (status.Status == JobStatus.Completed && status.Result is not null)
            {
                // The result is stored as PdfSplitResult in JobStatusResponse
                // We need to return it as TResult
                if (status.Result is TResult typedResult)
                {
                    return typedResult;
                }

                // This shouldn't happen if types are used correctly
                throw new JobException("Job completed but result type mismatch", status.JobId);
            }

            if (status.Status == JobStatus.Failed)
            {
                throw new JobException(status.Error ?? "Job failed", status.JobId);
            }

            attempts++;
            await Task.Delay(_pollInterval, cancellationToken).ConfigureAwait(false);
        }

        throw new JobException("Job polling timeout exceeded");
    }

    private void LogJobStatus(JobStatusResponse status, ref string? lastLoggedStatus, ref int? lastLoggedProgress)
    {
        if (_logger is null)
        {
            return;
        }

        var currentStatus = status.Status.ToString().ToLowerInvariant();
        var currentProgress = status.Progress;

        // Only log if status or progress changed
        var statusChanged = currentStatus != lastLoggedStatus;
        var progressChanged = currentProgress != lastLoggedProgress && currentProgress.HasValue;

        if (!statusChanged && !progressChanged)
        {
            return;
        }

        var progressStr = currentProgress.HasValue ? $" ({currentProgress}%)" : string.Empty;
        _logger.Log($"[Renamed] Job {status.JobId}: {currentStatus}{progressStr}");

        lastLoggedStatus = currentStatus;
        lastLoggedProgress = currentProgress;
    }
}

File: sdks/ruby/lib/renamed/async_job.rb
# frozen_string_literal: true

require_relative "models"
require_relative "errors"

module Renamed
  # Async job handle for long-running operations like PDF split.
  class AsyncJob
    DEFAULT_POLL_INTERVAL = 2.0
    MAX_POLL_ATTEMPTS = 150 # 5 minutes at 2s intervals

    attr_reader :status_url

    # @param client [Client] The Renamed client instance
    # @param status_url [String] URL to poll for job status
    # @param poll_interval [Float] Interval between status checks in seconds
    # @param max_attempts [Integer] Maximum number of polling attempts
    def initialize(client, status_url, poll_interval: DEFAULT_POLL_INTERVAL, max_attempts: MAX_POLL_ATTEMPTS)
      @client = client
      @status_url = status_url
      @poll_interval = poll_interval
      @max_attempts = max_attempts
    end

    # Get current job status.
    #
    # @return [JobStatusResponse] Current job status
    def status
      response = @client.send(:request, :get, @status_url)
      JobStatusResponse.from_response(response)
    end

    # Wait for job completion, polling at regular intervals.
    #
    # @yield [JobStatusResponse] Called with status updates during polling
    # @return [PdfSplitResult] The completed job result
    # @raise [JobError] If the job fails or times out
    #
    # @example Wait for job completion
    #   job = client.pdf_split("multi-page.pdf", mode: "auto")
    #   result = job.wait { |status| puts "Progress: #{status.progress}%" }
    #   result.documents.each do |doc|
    #     puts "#{doc.filename}: #{doc.download_url}"
    #   end
    def wait
      attempts = 0

      while attempts < @max_attempts
        current_status = status

        log_job_status(current_status)
        yield current_status if block_given?

        if current_status.status == "completed" && current_status.result
          return current_status.result
        end

        if current_status.status == "failed"
          raise JobError.new(current_status.error || "Job failed", job_id: current_status.job_id)
        end

        attempts += 1
        sleep(@poll_interval)
      end

      raise JobError.new("Job polling timeout exceeded")
    end

    private

    def log_job_status(job_status)
      @client.send(:log_job_status, job_status.job_id, job_status.status, job_status.progress)
    end
  end
end
